### PR TITLE
split setup_checkpointing

### PR DESCRIPTION
**Pull request by TahaZahid05**
### Overview
This pull request refactors the monolithic checkpoint setup function `setup_checkpointing` in `trainite/shared/utils.py` into two focused helper functions:
1. `setup_training_checkpointing` (for saving `last.pt` checkpoint on training epoch completion).
2. `setup_best_model_checkpoint` (for saving `best.pt` checkpoint on validation completion based on loss).

### CODE REVIEW

File: trainite/shared/utils.py [hunk 1]
```diff
@@ -255,14 +255,28 @@ def attach_early_stopping(
...
+
+def setup_best_model_checkpoint(
+    engine: Engine,
+    val_evaluator: Engine,
+    to_save: dict[str, Any],
+    run_dir: Path,
+) -> Checkpoint:
     def score_function(engine_val):
         loss = engine_val.state.metrics["loss"]
         return -loss
```
Comment (aaishwarymishra) [inline trainite/shared/utils.py]: Maybe its better, to take score function as argument too, we can define one at trainer. But the thing would be where will it pass cleanely.

Comment (TahaZahid05) [reply trainite/shared/utils.py]: we can add as optional args to `setup_best_model_checkpoint`?

File: trainite/trainers/decoder_trainer.py [hunk 2]
```diff
@@ -148,6 +149,36 @@ def __init__(self, config: ProjectConfig) -> None:
...
+                best_checkpoint = self.checkpointers.get("checkpoint_best")
+                checkpoint_path = best_checkpoint.last_checkpoint if best_checkpoint else None
+                if checkpoint_path and Path(checkpoint_path).exists():
+                    self.logger.info(f"Uploading new best model artifact to W&B: {checkpoint_path}")
+                    artifact = self.exp_logger.Artifact(
+                        name=f"{config.output.run_name}-model".replace("/", "-"), type="model"
+                    )
+                    artifact.add_file(str(checkpoint_path))
+                    self.exp_logger.log_artifact(artifact)
+
```
Comment (aaishwarymishra) [inline trainite/trainers/decoder_trainer.py]: we should only upload when the best model changes, it will slow the training, and how you make sure it uploads the bes model, the case could be the last checkpoint triggered it.

File: trainite/trainers/decoder_trainer.py [hunk 3]
```diff
@@ -148,6 +149,36 @@ def __init__(self, config: ProjectConfig) -> None:
...
+                last_checkpoint = self.checkpointers.get("checkpoint_last")
+                checkpoint_path = last_checkpoint.last_checkpoint if last_checkpoint else None
+                if checkpoint_path and Path(checkpoint_path).exists():
+                    self.logger.info(f"Uploading last checkpoint artifact to W&B: {checkpoint_path}")
+                    artifact = self.exp_logger.Artifact(
+                        name=f"{config.output.run_name}-checkpoint".replace("/", "-"), type="checkpoint"
+                    )
+                    artifact.add_file(str(checkpoint_path))
+                    self.exp_logger.log_artifact(artifact)
+
```
Comment (aaishwarymishra) [inline trainite/trainers/decoder_trainer.py]: same here

File: trainite/trainers/decoder_trainer.py
Comment (aaishwarymishra) [inline trainite/trainers/decoder_trainer.py:1]: I think this is polluting the init method

File: trainite/shared/utils.py (hunk 1 above)
Comment (aaishwarymishra) [reply trainite/shared/utils.py]: I think it should take it as required. Also I think we sholuld first merge the refactored checkpointing code, the artifact uploading can be a separate pr itself.

Comment (TahaZahid05) [reply trainite/shared/utils.py]: Why not have a default score_function and score_name and then accept if user want to override. I think its better. Both refactored, and artifact were related to checkpointing so i think it fits in one pr.

File: trainite/trainers/decoder_trainer.py (hunk 2 above)
Comment (TahaZahid05) [reply trainite/trainers/decoder_trainer.py]: It does only upload when best model changes? Best and last checkpoint are scope of different engines?

File: trainite/trainers/decoder_trainer.py (hunk 3 above)
Comment (TahaZahid05) [reply trainite/trainers/decoder_trainer.py]: same answer as above. they both have different scopes. and the event is only fired when a new model is written to disk

File: trainite/trainers/decoder_trainer.py
Comment (TahaZahid05) [reply trainite/trainers/decoder_trainer.py:1]: we can create a function instead once we are clear on the other issues with the code.

File: trainite/shared/utils.py (hunk 1 above)
Comment (aaishwarymishra) [reply trainite/shared/utils.py]: Not sure , its cleaner to send score function themselves.

File: trainite/trainers/decoder_trainer.py (hunk 2 above)
Comment (aaishwarymishra) [reply trainite/trainers/decoder_trainer.py]: hmm, my bad I got confused

File: trainite/trainers/decoder_trainer.py
Comment (aaishwarymishra) [reply trainite/trainers/decoder_trainer.py:1]: okay

File: trainite/shared/utils.py (hunk 1 above)
Comment (TahaZahid05) [reply trainite/shared/utils.py]: ok

File: trainite/trainers/decoder_trainer.py [hunk 4]
```diff
@@ -286,6 +254,37 @@ def _run_evaluations(self, engine: Engine) -> None:
             val_metrics["token_accuracy"],
         )
 
+    def _setup_wandb_checkpoint_uploads(self) -> None:
+        """Register events and upload handlers to automatically log checkpoints to W&B in real-time."""
+        self.val_evaluator.register_events(*CheckpointEvents)
+        self.trainer.register_events(*CheckpointEvents)
+
+        # 1. Upload best model artifact immediately when a better one is saved
+        @self.val_evaluator.on(CheckpointEvents.SAVED_CHECKPOINT)
```
Comment (aaishwarymishra) [inline trainite/trainers/decoder_trainer.py]: should use attach_handler method for consistency

File: trainite/shared/utils.py [hunk 5]
```diff
@@ -276,17 +272,15 @@ def setup_best_model_checkpoint(
     val_evaluator: Engine,
     to_save: dict[str, Any],
     run_dir: Path,
+    score_function: Callable[[Engine], float],
```
Comment (aaishwarymishra) [inline trainite/shared/utils.py]: I think callable is enough to keep it simple for the user

File: trainite/trainers/decoder_trainer.py [hunk 6]
```diff
@@ -112,24 +96,33 @@ def __init__(self, config: ProjectConfig) -> None:
...
-        attach_early_stopping(self.val_evaluator, self.engine, config.trainer.early_stopping_patience)
+        attach_early_stopping(self.val_evaluator, self.trainer, self.trainer_config.early_stopping_patience)
+
+        # Define validation scoring function for best model
+        def score_function(engine_val):
+            loss = engine_val.state.metrics["loss"]
+            return -loss
 
         # Attach checkpointing
         self.checkpointers = {
```
Comment (aaishwarymishra) [inline trainite/trainers/decoder_trainer.py]: its cleaner to define self.checkpointers as empty dict first, and then individually adding checkpoints

**Comment by TahaZahid05** [comment]
@aaishwarymishra done.

File: trainite/trainers/decoder_trainer.py [hunk 7]
```diff
@@ -234,6 +253,36 @@ def _run_evaluations(self, engine: Engine) -> None:
             val_metrics["token_accuracy"],
         )
 
+    def _setup_wandb_checkpoint_uploads(self) -> None:
```
Comment (aaishwarymishra) [inline trainite/trainers/decoder_trainer.py]: I think its more suited for utils.py

**Comment by TahaZahid05** [comment]
@aaishwarymishra done.

File: trainite/shared/utils.py [hunk 8]
```diff
@@ -389,3 +373,38 @@ def setup_console_logger(
...
+        if checkpoint_path and Path(checkpoint_path).exists():
+            logger.info(f"Uploading new best model artifact to W&B: {checkpoint_path}")
+            artifact = exp_logger.Artifact(name=f"{run_name}-model".replace("/", "-"), type="model")
+            artifact.add_file(str(checkpoint_path))
+            exp_logger.log_artifact(artifact)
+
+    def upload_last_checkpoint_artifact(engine):
+        last_checkpoint = checkpointers.get("checkpoint_last")
+        checkpoint_path = last_checkpoint.last_checkpoint if last_checkpoint else None
+        if checkpoint_path and Path(checkpoint_path).exists():
```
Comment (aaishwarymishra) [inline trainite/shared/utils.py]: isnt this check redundant if we using checkpoint saved event. Also can you check with a wandb key if it actually works

Comment (TahaZahid05) [reply trainite/shared/utils.py]: true, i was being a bit more defensive to ensure file exists even though event fires. i will change ig

**Review by aaishwarymishra — APPROVED** [review]

**Comment by aaishwarymishra** [comment]
I would suggest running an experiment to see if artifacts gets uploaded before merging.

**Comment by TahaZahid05** [comment]
@aaishwarymishra sorry didn't mention it, but I did run experiment and artifacts are getting uploaded correctly.
